File: services/self-service/src/main/resources/webapp/src/app/core/services/project.service.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { map, catchError } from 'rxjs/operators';

import { ApplicationServiceFacade } from './applicationServiceFacade.service';
import { ErrorUtils } from '../util';

@Injectable()
export class ProjectService {
  constructor(private applicationServiceFacade: ApplicationServiceFacade) { }

  public createProject(data): Observable<{}> {
    return this.applicationServiceFacade
      .buildCreateProject(data)
      .pipe(
        map(response => response),
        catchError(ErrorUtils.handleServiceError));
  }

  public updateProject(data): Observable<{}> {
    return this.applicationServiceFacade
      .buildUpdateProject(data)
      .pipe(
        map(response => response),
        catchError(ErrorUtils.handleServiceError));
  }

  public getProjectsList(): Observable<{}> {
    return this.applicationServiceFacade
      .buildGetProjectsList()
      .pipe(
        map(response => response),
        catchError(ErrorUtils.handleServiceError));
  }

  public getUserProjectsList(isActive?): Observable<{}> {
    const params = isActive ? '/me?active=true' : '';
    return this.applicationServiceFacade
      .buildGetUserProjectsList(params)
      .pipe(
        map(response => response),
        catchError(ErrorUtils.handleServiceError));
  }

  public toggleProjectStatus(data, action): Observable<{}> {
    const url = `/${action}`;
    return this.applicationServiceFacade
      .buildToggleProjectStatus(url, data)
      .pipe(
        map(response => response),
        catchError(ErrorUtils.handleServiceError));
  }

  public updateProjectsBudget(data): Observable<{}> {
    const url = '/budget';
    return this.applicationServiceFacade
      .buildUpdateProjectsBudget(url, data)
      .pipe(
        map(response => response),
        catchError(ErrorUtils.handleServiceError));
  }
}

File: services/self-service/src/main/java/com/epam/datalab/backendapi/service/ProjectService.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.epam.datalab.backendapi.service;

import com.epam.datalab.auth.UserInfo;
import com.epam.datalab.backendapi.domain.ProjectDTO;
import com.epam.datalab.backendapi.domain.UpdateProjectBudgetDTO;
import com.epam.datalab.backendapi.domain.UpdateProjectDTO;
import com.epam.datalab.dto.UserInstanceStatus;

import java.util.List;

public interface ProjectService {
	List<ProjectDTO> getProjects();

	List<ProjectDTO> getProjects(UserInfo user);

	List<ProjectDTO> getUserProjects(UserInfo userInfo, boolean active);

	List<ProjectDTO> getProjectsByEndpoint(String endpointName);

	void create(UserInfo userInfo, ProjectDTO projectDTO, String resourceName);

	void recreate(UserInfo userInfo, String endpoint, String projectName);

	ProjectDTO get(String name);

	void terminateEndpoint(UserInfo userInfo, String endpoint, String name);

	void terminateEndpoint(UserInfo userInfo, List<String> endpoints, String name);

	void start(UserInfo userInfo, String endpoint, String name);

	void start(UserInfo userInfo, List<String> endpoints, String name);

    void stop(UserInfo userInfo, String endpoint, String name, String auditInfo);

	void stopWithResources(UserInfo userInfo, List<String> endpoints, String projectName);

	void update(UserInfo userInfo, UpdateProjectDTO projectDTO, String projectName);

	void updateBudget(UserInfo userInfo, List<UpdateProjectBudgetDTO> projects);

	boolean isAnyProjectAssigned(UserInfo userInfo);

	boolean checkExploratoriesAndComputationalProgress(String projectName, List<String> endpoints);

	void updateAfterStatusCheck(UserInfo userInfo, String project, String endpoint, String instanceID, UserInstanceStatus status, String auditInfo);
}
